blockly/blocks/component_variable.js

```diff
@@ -0,0 +1,172 @@
+/**
+ * @license
+ * Visual Blocks Editor
+ *
+ * Copyright 2012 Google Inc.
+ * https://developers.google.com/blockly/
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Variable input field of a specific component type.
+ * @author fraser@google.com (Neil Fraser)
+ */
+'use strict';
+
+goog.provide('Blockly.Blocks.ComponentFieldVariable');
+
+goog.require('Blockly.Blocks');
+//goog.require('Blockly.FieldVariable');
+
+
+/**
+ * Class for a variable's dropdown field.
+ * @param {?string} varname The default name for the variable.  If null,
+ *     a unique variable name will be generated.
+ * @param {Function=} opt_validator A function that is executed when a new
+ *     option is selected.  Its sole argument is the new option value.
+ * @extends {Blockly.FieldVariable}
+ * @constructor
+ */
+Blockly.Blocks.ComponentFieldVariable = function(varname, component_type, opt_validator) {
+  this.component_type = component_type
+  //override the dropdownCreate to use for this field
+  this.dropdownCreate = Blockly.Blocks.ComponentFieldVariable.dropdownCreateComponents;
```

Comment:
Best solution would be some more rework in the parent class, but this was the way that needs least changes in parent class of core blockly. It works due to the changes in the parent class